### ISSUE TITLE
Configuration is not thread-safe

### ISSUE THREAD

**Issue by jnyrup** [fluentassertions/fluentassertions#664]
When executing all tests in VS with "Run tests in parallel" enabled, sometimes some tests fails.

The stack trace:
```
Expected a <Xunit.Sdk.XunitException> to be thrown, but found a <System.InvalidOperationException>: System.InvalidOperationException with message "Nullable object must have a value."
     at System.ThrowHelper.ThrowInvalidOperationException(ExceptionResource resource)
     at System.Nullable`1.get_Value()
     at FluentAssertions.Common.Configuration.get_ValueFormatterDetectionMode() in C:\git\fluentassertions\Src\FluentAssertions\Common\Configuration.cs:line 43
     at FluentAssertions.Formatting.AttributeBasedFormatter.get_IsScanningEnabled() in C:\git\fluentassertions\Src\FluentAssertions\Formatting\AttributeBasedFormatter.cs:line 34
     at FluentAssertions.Formatting.AttributeBasedFormatter.CanHandle(Object value) in C:\git\fluentassertions\Src\FluentAssertions\Formatting\AttributeBasedFormatter.cs:line 29
     at FluentAssertions.Formatting.Formatter.<>c__DisplayClass8_0.<Format>b__0(IValueFormatter f) in C:\git\fluentassertions\Src\FluentAssertions\Formatting\Formatter.cs:line 138
     at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
     at System.Linq.Enumerable.First[TSource](IEnumerable`1 source, Func`2 predicate)
     at FluentAssertions.Formatting.Formatter.Format(Object value, FormattingContext context, FormatChild formatChild) in C:\git\fluentassertions\Src\FluentAssertions\Formatting\Formatter.cs:line 138
     at FluentAssertions.Formatting.Formatter.ToString(Object value, Boolean useLineBreaks) in C:\git\fluentassertions\Src\FluentAssertions\Formatting\Formatter.cs:line 91
     at FluentAssertions.Execution.MessageBuilder.<FormatArgumentPlaceholders>b__7_0(Object a) in C:\git\fluentassertions\Src\FluentAssertions\Execution\MessageBuilder.cs:line 62
     at System.Linq.Enumerable.SelectArrayIterator`2.ToArray()
     at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
     at FluentAssertions.Execution.MessageBuilder.FormatArgumentPlaceholders(String failureMessage, Object[] failureArgs) in C:\git\fluentassertions\Src\FluentAssertions\Execution\MessageBuilder.cs:line 62
     at FluentAssertions.Execution.MessageBuilder.Build(String message, Object[] messageArgs, String reason, ContextDataItems contextData) in C:\git\fluentassertions\Src\FluentAssertions\Execution\MessageBuilder.cs:line 41
     at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args) in C:\git\fluentassertions\Src\FluentAssertions\Execution\AssertionScope.cs:line 190
     at FluentAssertions.Primitives.StringEqualityValidator.ValidateAgainstMismatch() in C:\git\fluentassertions\Src\FluentAssertions\Primitives\StringEqualityValidator.cs:line 44
     at FluentAssertions.Primitives.StringValidator.Validate() in C:\git\fluentassertions\Src\FluentAssertions\Primitives\StringValidator.cs:line 43
     at FluentAssertions.Primitives.StringAssertions.Be(String expected, String because, Object[] becauseArgs) in C:\git\fluentassertions\Src\FluentAssertions\Primitives\StringAssertions.cs:line 39
     at FluentAssertions.Specs.StringAssertionSpecs.<>c.<When_two_strings_differ_and_one_of_them_is_multiline_it_should_display_both_strings_on_separate_line>b__11_0() in C:\git\fluentassertions\Tests\Shared.Specs\StringAssertionSpecs.cs:line 184
     at FluentAssertions.Specialized.ActionAssertions.InvokeSubjectWithInterception() in C:\git\fluentassertions\Src\FluentAssertions\Specialized\ActionAssertions.cs:line 112
```

Here's a test to reproduce it.
```c#
[Fact]
public void When_running_tests_in_parallel_it_should_still_work()
{
    Parallel.For(
        0,
        10000,
        new ParallelOptions
        {
            MaxDegreeOfParallelism = 8
        },
        e =>
        {
            Configuration.Current.ValueFormatterAssembly = "";
            var mode = Configuration.Current.ValueFormatterDetectionMode;
        }
    );
}
```

So the problem seems to be a race condition `set_ValueFormatterAssembly()` setting `valueFormatterDetectionMode = null` and `Nullable.get_ValueFormatterDetectionMode()` initializing `valueFormatterDetectionMode`.

**Comment by jnyrup** [comment]
Just a note:
When upgrading xunit from 2.3.0-beta5 to 2.3.0, the regular `build.ps1` failed _every_ time on `When_a_custom_formatter_exists_in_any_loaded_assembly_it_should_override_the_default_formatters`.

**Comment by jnyrup** [comment]
Fixed in #714
